[DOC] Modular monitoring appendix (#1736)
strimzi/strimzi-kafka-operator: documentation/book/ref-metrics-alertmanager-examples.adoc

// This assembly is included in the following assemblies:
//
// assembly-metrics-prometheus-alertmanager.adoc
[id='ref-metrics-alertmanager-examples-{context}']

= Alerting rule examples

Example alerting rules for Kafka and Zookeeper metrics are provided with {ProductName} for use in a xref:proc-metrics-deploying-prometheus-{context}[Prometheus deployment].

The rules are defined in a `PrometheusRule` resource in the link:https://raw.githubusercontent.com/strimzi/strimzi-kafka-operator/{GithubVersion}/metrics/examples/prometheus/install/prometheus-rules.yaml[`prometheus-rules.yaml`^] file

.Kafka alerting rules

> * `UnderReplicatedPartitions`: gives the number of partitions for which the current broker is the lead replica but the follower replicas are behind.

Reviewer: We should define "behind". @ppatierno can you confirm that the following is correct.
```suggestion
* `UnderReplicatedPartitions`: gives the number of partitions for which the current broker is the lead replica but which have fewer replicas than the `min.insync.replicas` configured for their topic.
```

> This metric provides insights about offline brokers that host the follower replicas.

Reviewer: They're not necessarily offline. They may have been off line and are still catching up. The interbroker replication may be throttled, etc.
```suggestion
This metric provides insights about brokers that host the follower replicas. 
Those followers are not keeping up with the leader.
Reasons for this could include being (or having been) offline, and over-throttled interbroker replication.
```

An alert is raised when this value is greater than zero, providing information on the under-replicated partitions for each broker.

* `AbnormalControllerState`: indicates whether the current broker is the controller for the cluster.
The metric can be 0 or 1.
During the life of a cluster, only one broker should be the controller and the cluster always needs to have an active controller.
Having two or more brokers saying that they are controllers indicates a problem.

> An alert is raised when the sum of all the values for this metric on all brokers is not equal to 1, meaning that there is no active controller (the sum is 0) or more than one controller (the sum is greater than 1).

Reviewer: @ppatierno is it raised immediately? What happens during controller election?

Reviewer 2: The alerts should be raised if the condition persists for some time which is configurable IIRC. Right @stanlyDoge ?

Reviewer 3: Yes, when the condition is fulfilled there is some time bore the alert is fired. It is set by option 
`for: 10s`

Author: Updated to mention the `for` property

* `UnderMinIsrPartitionCount`: indicates that the minimum number of in-sync replicas (ISRs) for a lead Kafka broker, specified using `min.insync.replicas`, that must acknowledge a write operation has not been reached.
The metric defines the number of partitions that the broker leads for which the in-sync replicas count is less than the minimum in-sync.
An alert is raised when this value is greater than zero, providing information on the partition count for each broker that did not achieve the minimum number of acknowledgments.

* `OfflineLogDirectoryCount`: indicates the number of log directories which are offline (for example, due to a hardware failure) so that the broker cannot store incoming messages anymore.
An alert is raised when this value is greater than zero, providing information on the number of offline log directories for each broker.

* `KafkaRunningOutOfSpace`: indicates the remaining amount of disk space that can be used for writing data.

> An alert is raised when this value is lower than 5GiB, providing information on the disk that is running out of space for each persistent volume claim.

Reviewer: 5GiB doesn't sound like very much. How can the user configure this?

Author: @ppatierno or @stanlyDoge - can we check on this alert?

Reviewer 3: User can redeploy `prometheus-rules.yaml` file with his own values. What value should be set as "default"?

Author: I've added that teh threshold value can be changed in `prometheus-rules.yaml`

> NOTE: The availability of the `KafkaRunningOutOfSpace` metric and alert is dependent on the version of {ProductPlatformName}.

Reviewer: I don't think this is sufficient. How does the user know whether they should have it, or what version they need to use to have it?

Author: @ppatierno or @stanlyDoge - can we give some clearer guidance on the availability of this metric?

Reviewer 2: @stanlyDoge I think that you added this one doing some tests on Kubernetes and OpenShift. Can you give guidance about versions?

Reviewer 4: I do not think this is dependent on `version of {ProductPlatformName}`. This will be dependent on `{ProductPlatformName} configuration and used storage implementation`.

Author: The availability of the `ZookeeperRunningOutOfSpace` metric and alert is dependent on the {ProductPlatformName} configuration and storage implementation used. Storage implementations for certain platforms may not be able to supply the information on available space for the metric to be able to provide an alert.

.Zookeeper alerting rules

* `AvgRequestLatency`: indicates the amount of time it takes for the server to respond to a client request.

> An alert is raised when this value is greater than 10 (ticks), providing the actual value of the average request latency for each server.

Reviewer: This doesn't make sense unless we've explained what a tick is.

Author: NOTE: A tick is a basic Zookeeper time unit, which is measured in milliseconds and configured using the `tickTime` parameter. For example, if Zookeeper `tickTime=3000`, 3 ticks (3 x 3000) equals 9000 milliseconds.

* `OutstandingRequests`: indicates the number of queued requests in the server.
This value goes up when the server receives more requests than it can process.

> An alert is raised when this value is greater than 10 (ticks), providing the actual number of outstanding requests for each server.

Reviewer: How can this be in ticks? Surely it's a count of the number of outstanding requests?

Author: @ppatierno or @stanlyDoge  - can we check on this alert?

Reviewer 3: @ppatierno ^

Reviewer 2: I guess that's a typo, "ticks" doesn't make sense here.

// This assembly is included in the following assemblies:
//
// assembly-metrics-prometheus-alertmanager.adoc
[id='ref-metrics-alertmanager-examples-{context}']

= Alerting rule examples

Example alerting rules for Kafka and Zookeeper metrics are provided with {ProductName} for use in a xref:proc-metrics-deploying-prometheus-{context}[Prometheus deployment].

The rules are defined in a `PrometheusRule` resource in the link:https://raw.githubusercontent.com/strimzi/strimzi-kafka-operator/{GithubVersion}/metrics/examples/prometheus/install/prometheus-rules.yaml[`prometheus-rules.yaml`^] file.

General points about the alerting rule definitions:

* A `for` property is used with the rules to determine the period of time a condition must persist before an alert is triggered.

> * A tick is a basic Zookeeper time unit, which is measured in milliseconds and configured using the `tickTime` parameter. For example, if Zookeeper `tickTime=3000`, 3 ticks (3 x 3000) equals 9000 milliseconds.

Reviewer: ```suggestion
* A tick is a basic Zookeeper time unit, which is measured in milliseconds and configured using the `tickTime` parameter of `Kafka.spec.zookeeper.config`. For example, if Zookeeper `tickTime=3000`, 3 ticks (3 x 3000) equals 9000 milliseconds.
```
